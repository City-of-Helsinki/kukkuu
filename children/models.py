from hashlib import sha256
from typing import Iterable, Optional

from django.conf import settings
from django.core.validators import RegexValidator
from django.db import models, transaction
from django.db.models import Q
from django.utils import timezone
from django.utils.translation import gettext_lazy as _

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from languages.models import Language
from users.models import Guardian


class ChildQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardians__user=user) | Q(project__in=user.administered_projects)
        ).distinct()

    def user_can_update(self, user):
        return self.filter(guardians__user=user)

    def user_can_delete(self, user):
        return self.filter(guardians__user=user)

    @transaction.atomic()
    def delete(self):
        for child in self:
            child.delete()


postal_code_validator = RegexValidator(
    regex=r"^\d{5}$", message="Postal code must be 5 digits", code="invalid_postal_code"
)


class Child(UUIDPrimaryKeyModel, TimestampedModel):
    name = models.CharField(verbose_name=_("name"), max_length=64, blank=True)
    birthyear = models.DateField(verbose_name=_("birthyear"))
    postal_code = models.CharField(
        verbose_name=_("postal code"),
        max_length=5,
        validators=[postal_code_validator],
    )
    guardians = models.ManyToManyField(
        Guardian,
        verbose_name=_("guardians"),
        related_name="children",
        through="children.Relationship",
        blank=True,
    )
    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="children",
        on_delete=models.PROTECT,
    )
    languages_spoken_at_home = models.ManyToManyField(
        Language,
        verbose_name=_("languages spoken at home"),
        related_name="children",
        blank=True,
    )

    objects = ChildQuerySet.as_manager()

    class Meta:
        verbose_name = _("child")
        verbose_name_plural = _("children")
        ordering = ["birthyear", "name"]

    def __str__(self):
        return f"{self.name} ({self.birthyear})"

    @transaction.atomic()
    def delete(self, *args, **kwargs):
        self.enrolments.upcoming().delete()
        return super().delete(*args, **kwargs)

    def can_user_administer(self, user):
        return user.can_administer_project(self.project)

    def get_enrolment_count(self, year: Optional[int] = None, past=False):
        from events.models import Event  # noqa

        if year and past:
            raise ValueError("Cannot use year and past arguments at the same time.")
        now = timezone.now()
        year = year or now.year

        occurrence_filters = Q(time__year=year)
        # For external ticket system events published_at field is used to determine the
        # event's year. That is obviously not a perfect solution, but the best we can
        # do with the current data, and most probably good enough.
        ticket_system_filters = Q(event__published_at__year=year)
        if past:
            occurrence_filters &= Q(time__lt=now)
            ticket_system_filters &= Q(event__published_at__lt=now)
        return (
            self.occurrences.filter(occurrence_filters).count()
            + self.ticket_system_passwords.filter(ticket_system_filters)
            .distinct()
            .count()
        )

    @staticmethod
    def _hash(values: Iterable[str]) -> str:
        return sha256(",".join(values).encode("utf8")).hexdigest()

    @property
    def _salt_birthyear_postal_code_guardian_emails_list(self) -> list[str]:
        """
        List of salt, child's birthyear, postal code and guardians' emails.
        """
        return [
            settings.KUKKUU_HASHID_SALT,
            self.birthyear.isoformat(),
            self.postal_code,
            *sorted(guardian.email for guardian in self.guardians.all()),
        ]

    @property
    def birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's birthyear, postal code and guardians' emails.
        """
        return self._hash(self._salt_birthyear_postal_code_guardian_emails_list)

    @property
    def name_birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's name, birthyear, postal
        code and guardians' emails.
        """
        return self._hash(
            [self.name] + self._salt_birthyear_postal_code_guardian_emails_list
        )


class RelationshipQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardian__user=user) | Q(child__project__in=user.administered_projects)
        ).distinct()


class Relationship(models.Model):
    PARENT = "parent"  # In Finnish: Vanhempi
    OTHER_GUARDIAN = "other_guardian"  # In Finnish: Muu huoltaja
    OTHER_RELATION = "other_relation"  # In Finnish: Muu l√§heinen
    ADVOCATE = "advocate"  # In Finnish: Virallinen puolesta-asioija
    TYPE_CHOICES = (
        (PARENT, _("Parent")),
        (OTHER_GUARDIAN, _("Other guardian")),
        (OTHER_RELATION, _("Other relation")),
        (ADVOCATE, _("Advocate")),
    )

    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    child = models.ForeignKey(
        Child,
        verbose_name=_("child"),
        related_name="relationships",
        on_delete=models.CASCADE,
    )
    guardian = models.ForeignKey(
        Guardian,
        verbose_name=_("guardian"),
        on_delete=models.CASCADE,
        related_name="relationships",
    )
    type = models.CharField(
        verbose_name=_("type"),
        choices=TYPE_CHOICES,
        max_length=64,
        null=True,
        blank=True,
    )

    objects = RelationshipQuerySet.as_manager()

    class Meta:
        verbose_name = _("relationship")
        verbose_name_plural = _("relationships")
        ordering = [
            "guardian",
            "child",
        ]
