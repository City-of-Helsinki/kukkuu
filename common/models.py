import uuid

from django.db import models
from django.utils.translation import ugettext_lazy as _
from parler.managers import TranslatableQuerySet as ParlerTranslatableQuerySet
from parler.models import TranslatableModel as ParlerTranslatableModel


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated_at"), auto_now=True)

    class Meta:
        abstract = True


class UUIDPrimaryKeyModel(models.Model):
    id = models.UUIDField(
        verbose_name=_("UUID"), primary_key=True, default=uuid.uuid4, editable=False
    )

    class Meta:
        abstract = True


class TranslatableQuerySet(ParlerTranslatableQuerySet):
    def create_translatable_object(self, **kwargs):
        translations = kwargs.pop("translations")
        obj = self.create(**kwargs)
        for translation in translations:
            language_code = translation.pop("language_code")
            obj.create_translation(language_code=language_code, **translation)
        return obj


class TranslatableModel(ParlerTranslatableModel):
    objects = TranslatableQuerySet.as_manager()

    class Meta:
        abstract = True
