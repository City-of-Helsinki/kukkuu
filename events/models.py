from django.conf import settings
from django.db import models
from django.db.models import Q
from django.utils import timezone
from django.utils.timezone import now
from django.utils.translation import ugettext_lazy as _
from parler.models import TranslatedFields

from children.models import Child
from common.models import TimestampedModel, TranslatableModel, TranslatableQuerySet
from events.consts import NotificationType
from events.utils import send_event_notifications_to_guardians
from venues.models import Venue


# This need to be inherited from TranslatableQuerySet instead of default model.QuerySet
class EventQueryset(TranslatableQuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(project__users=user) | Q(published_at__isnull=False)
        ).distinct()

    def published(self):
        return self.filter(published_at__isnull=False)


class Event(TimestampedModel, TranslatableModel):
    CHILD_AND_GUARDIAN = "child_and_guardian"
    CHILD_AND_1_OR_2_GUARDIANS = "child_and_1_or_2_guardians"
    FAMILY = "family"
    PARTICIPANTS_PER_INVITE_CHOICES = (
        (CHILD_AND_GUARDIAN, _("Child and guardian")),
        (CHILD_AND_1_OR_2_GUARDIANS, _("Child and 1-2 guardians")),
        (FAMILY, _("Family")),
    )

    translations = TranslatedFields(
        name=models.CharField(verbose_name=_("name"), max_length=255, blank=True),
        short_description=models.TextField(
            verbose_name=_("short description"), blank=True
        ),
        description=models.TextField(verbose_name=_("description"), blank=True),
        image_alt_text=models.CharField(
            verbose_name=_("image alt text"), blank=True, max_length=255
        ),
    )
    image = models.ImageField(blank=True, verbose_name=_("image"))
    participants_per_invite = models.CharField(
        max_length=255,
        choices=PARTICIPANTS_PER_INVITE_CHOICES,
        verbose_name=_("participants per invite"),
    )
    duration = models.PositiveSmallIntegerField(
        verbose_name=_("duration"), blank=True, null=True, help_text=_("In minutes")
    )
    capacity_per_occurrence = models.PositiveSmallIntegerField(
        verbose_name=_("capacity per occurrence")
    )
    published_at = models.DateTimeField(
        blank=True, null=True, verbose_name=_("published at")
    )

    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="events",
        on_delete=models.CASCADE,
    )

    objects = EventQueryset.as_manager()

    class Meta:
        verbose_name = _("event")
        verbose_name_plural = _("events")

    def __str__(self):
        name = self.safe_translation_getter("name", super().__str__())
        published_text = _("published") if self.published_at else _("unpublished")
        return f"{name} ({self.pk}) ({self.project.year}) ({published_text})"

    def can_user_administer(self, user):
        return user.projects.filter(pk=self.project_id).exists()

    def publish(self):
        self.published_at = timezone.now()
        self.save()

        send_event_notifications_to_guardians(
            self,
            NotificationType.EVENT_PUBLISHED,
            self.project.children.prefetch_related("guardians"),
        )

    def is_published(self):
        return bool(self.published_at)


class OccurrenceQueryset(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(event__project__users=user) | Q(event__published_at__isnull=False)
        ).distinct()

    def delete(self, *args, **kwargs):
        for obj in self:
            obj.delete()


class Occurrence(TimestampedModel):
    time = models.DateTimeField(verbose_name=_("time"))
    event = models.ForeignKey(
        Event,
        verbose_name=_("event"),
        related_name="occurrences",
        on_delete=models.CASCADE,
    )
    venue = models.ForeignKey(
        Venue,
        verbose_name=_("venue"),
        related_name="occurrences",
        on_delete=models.CASCADE,
    )

    children = models.ManyToManyField(
        Child,
        verbose_name=_("children"),
        related_name="occurrences",
        through="events.Enrolment",
        blank=True,
    )
    occurrence_language = models.CharField(
        max_length=10,
        choices=settings.LANGUAGES,
        verbose_name=_("occurrence language"),
        default=settings.LANGUAGES[0][0],
    )
    objects = OccurrenceQueryset.as_manager()

    class Meta:
        verbose_name = _("occurrence")
        verbose_name_plural = _("occurrences")

    def __str__(self):
        return f"{self.time} ({self.pk})"

    def delete(self, *args, **kwargs):
        if self.time >= now():
            # this QS needs to be evaluated here, it would not work after the
            # occurrence has been deleted
            children = list(self.children.all())

            super().delete(*args, **kwargs)

            send_event_notifications_to_guardians(
                self.event,
                NotificationType.OCCURRENCE_CANCELLED,
                children,
                occurrence=self,
            )
        else:
            super().delete(*args, **kwargs)

    def get_enrolment_count(self):
        try:
            # try to use an annotated value
            return self.enrolment_count
        except AttributeError:
            return self.enrolments.count()

    def can_user_administer(self, user):
        # There shouldn't ever be a situation where event.project != venue.project
        # so we can just check one of them
        return user.projects.filter(pk=self.event.project.pk).exists()


class Enrolment(TimestampedModel):
    child = models.ForeignKey(
        Child,
        related_name="enrolments",
        on_delete=models.CASCADE,
        verbose_name=_("child"),
    )
    occurrence = models.ForeignKey(
        Occurrence,
        related_name="enrolments",
        on_delete=models.CASCADE,
        verbose_name=_("occurrence"),
    )
    attended = models.NullBooleanField(verbose_name=_("attended"))

    class Meta:
        verbose_name = _("enrolment")
        verbose_name_plural = _("enrolments")
        constraints = [
            models.UniqueConstraint(
                fields=["child", "occurrence"], name="unq_child_occurrence"
            )
        ]

    def __str__(self):
        return f"{self.pk} {self.child_id}"

    def save(self, *args, **kwargs):
        created = self.pk is None
        super().save(*args, **kwargs)

        if created:
            send_event_notifications_to_guardians(
                self.occurrence.event,
                NotificationType.OCCURRENCE_ENROLMENT,
                self.child,
                occurrence=self.occurrence,
            )

    def delete_and_send_notification(self):
        child = self.child
        occurrence = self.occurrence
        self.delete()

        send_event_notifications_to_guardians(
            occurrence.event,
            NotificationType.OCCURRENCE_UNENROLMENT,
            child,
            occurrence=occurrence,
        )

    def can_user_administer(self, user):
        return self.occurrence.can_user_administer(user)
