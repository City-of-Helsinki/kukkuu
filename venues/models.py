from django.db import models
from django.utils.translation import ugettext_lazy as _
from parler.models import TranslatedFields

from common.models import TimestampedModel, TranslatableModel


class Venue(TimestampedModel, TranslatableModel):
    translations = TranslatedFields(
        name=models.CharField(verbose_name=_("name"), max_length=255),
        description=models.TextField(verbose_name=_("description"), blank=True),
    )
    seat_count = models.PositiveIntegerField(verbose_name=_("seat count"))

    class Meta:
        verbose_name = _("venue")
        verbose_name_plural = _("venues")

    def __str__(self):
        return f"{self.pk} {self.name}"
