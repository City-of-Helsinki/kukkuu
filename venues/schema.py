import graphene
from django.apps import apps
from django.db import transaction
from django.db.models import Count
from django.utils.translation import get_language
from graphene import relay
from graphene_django import DjangoConnectionField, DjangoObjectType
from graphql_jwt.decorators import login_required, staff_member_required
from graphql_relay import from_global_id

from common.utils import update_object_with_translations
from kukkuu.exceptions import ObjectDoesNotExistError
from users.schema import LanguageEnum
from venues.models import Venue

VenueTranslation = apps.get_model("venues", "VenueTranslation")


class VenueTranslationType(DjangoObjectType):
    language_code = LanguageEnum(required=True)

    class Meta:
        model = VenueTranslation
        exclude = ("id", "master")


class VenueNode(DjangoObjectType):
    name = graphene.String()
    description = graphene.String()
    address = graphene.String()
    accessibility_info = graphene.String()
    arrival_instructions = graphene.String()
    additional_info = graphene.String()
    wc_and_facilities = graphene.String()
    www_url = graphene.String()

    class Meta:
        model = Venue
        interfaces = (relay.Node,)

    @classmethod
    @login_required
    # TODO: For now only logged in users can see venues
    def get_queryset(cls, queryset, info):
        lang = get_language()
        return queryset.order_by("-created_at").language(lang)

    @classmethod
    @login_required
    def get_node(cls, info, id):
        return super().get_node(info, id)

    def resolve_occurrences(self, info, **kwargs):
        return (
            self.occurrences.annotate(
                enrolments_count=Count("enrolments", distinct=True)
            )
            .select_related("event")
            .order_by("time")
        )


class VenueTranslationsInput(graphene.InputObjectType):
    name = graphene.String()
    description = graphene.String()
    language_code = LanguageEnum(required=True)
    address = graphene.String()
    accessibility_info = graphene.String()
    arrival_instructions = graphene.String()
    additional_info = graphene.String()
    www_url = graphene.String()
    wc_and_facilities = graphene.String()


class AddVenueMutation(graphene.relay.ClientIDMutation):
    class Input:
        translations = graphene.List(VenueTranslationsInput)

    venue = graphene.Field(VenueNode)

    @classmethod
    @staff_member_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        # TODO: Add validation
        venue = Venue.objects.create_translatable_object(**kwargs)
        return AddVenueMutation(venue=venue)


class UpdateVenueMutation(graphene.relay.ClientIDMutation):
    class Input:
        id = graphene.GlobalID(required=True)
        translations = graphene.List(VenueTranslationsInput)
        delete_translations = graphene.List(LanguageEnum)

    venue = graphene.Field(VenueNode)

    @classmethod
    @staff_member_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        # TODO: Add validation
        venue_global_id = kwargs.pop("id")
        try:
            venue = Venue.objects.get(pk=from_global_id(venue_global_id)[1])
            update_object_with_translations(venue, kwargs)
        except Venue.DoesNotExist as e:
            raise ObjectDoesNotExistError(e)
        return UpdateVenueMutation(venue=venue)


class DeleteVenueMutation(graphene.relay.ClientIDMutation):
    class Input:
        id = graphene.GlobalID()

    @classmethod
    @staff_member_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        # TODO: Validate data
        venue_id = from_global_id(kwargs["id"])[1]
        try:
            venue = Venue.objects.get(pk=venue_id)
            venue.delete()
        except Venue.DoesNotExist as e:
            raise ObjectDoesNotExistError(e)
        return DeleteVenueMutation()


class Query:
    venue = relay.Node.Field(VenueNode)
    venues = DjangoConnectionField(VenueNode)


class Mutation:
    add_venue = AddVenueMutation.Field()
    update_venue = UpdateVenueMutation.Field()
    delete_venue = DeleteVenueMutation.Field()
