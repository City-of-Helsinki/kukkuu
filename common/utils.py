import binascii
from copy import deepcopy
from functools import wraps

from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.db import transaction
from graphene import Node
from graphql.execution.base import ResolveInfo
from graphql_relay import from_global_id, to_global_id

from kukkuu import __version__
from kukkuu.exceptions import DataValidationError, ObjectDoesNotExistError
from kukkuu.settings import REVISION


def update_object(obj, data):
    if not data:
        return
    for k, v in data.items():
        if v is None and not obj.__class__._meta.get_field(k).null:
            raise DataValidationError(f"{k} cannot be null.")
        setattr(obj, k, v)
    obj.save()


@transaction.atomic
def update_object_with_translations(model, model_data):
    model_data = deepcopy(model_data)
    translations_input = model_data.pop("translations", None)
    if translations_input:
        model.create_or_update_translations(translations_input)
    update_object(model, model_data)


def get_api_version():
    return " | ".join((__version__, REVISION))


def get_global_id(obj):
    return to_global_id(f"{obj.__class__.__name__}Node", obj.pk)


def get_node_id_from_global_id(global_id, expected_node_name):
    if not global_id:
        return None
    try:
        name, id = from_global_id(global_id)
    except (
        binascii.Error,
        UnicodeDecodeError,
    ):  # invalid global ID
        return None
    return id if name == expected_node_name else None


def check_can_user_administer(obj, user):
    try:
        yes_we_can = obj.can_user_administer(user)
    except AttributeError:
        raise TypeError(
            f"{obj.__class__.__name__} model does not implement can_user_administer()."
        )
    if not yes_we_can:
        raise PermissionDenied()


def get_obj_from_global_id(info, global_id, expected_obj_type):
    obj = Node.get_node_from_global_id(info, global_id)
    if not obj or type(obj) is not expected_obj_type:
        raise ObjectDoesNotExistError(
            f"{expected_obj_type.__name__} with ID {global_id} does not exist."
        )
    return obj


def get_obj_if_user_can_administer(info, global_id, expected_obj_type):
    obj = get_obj_from_global_id(info, global_id, expected_obj_type)
    check_can_user_administer(obj, info.context.user)
    return obj


def context(f):
    def decorator(func):
        def wrapper(*args, **kwargs):
            info = next(arg for arg in args if isinstance(arg, ResolveInfo))
            return func(info.context, *args, **kwargs)

        return wrapper

    return decorator


def user_passes_test(test_func):
    def decorator(f):
        @wraps(f)
        @context(f)
        def wrapper(context, *args, **kwargs):
            if test_func(context.user):
                return f(*args, **kwargs)
            raise PermissionDenied("You do not have permission to perform this action")

        return wrapper

    return decorator


# copied from https://github.com/flavors/django-graphql-jwt/blob/704f24e7ebbea0b81015ef3c1f4a302e9d432ecf/graphql_jwt/decorators.py  # noqa
login_required = user_passes_test(lambda u: u.is_authenticated)

project_user_required = user_passes_test(
    lambda u: u.is_authenticated and u.administered_projects
)


def get_translations_dict(obj, field_name):
    """
    Returns a dict of translations for a given field of a model instance.

    :param obj: The model instance
    :param field_name: The name of the field
    :return: A dict with language codes as keys and translations as values
    """

    return {
        lang_code: getattr(
            obj.translations.filter(language_code=lang_code).first(), field_name, ""
        )
        for lang_code, _ in settings.LANGUAGES
    }
