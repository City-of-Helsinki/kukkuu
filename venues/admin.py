from django.contrib import admin
from parler.admin import TranslatableAdmin

from venues.models import Venue


@admin.register(Venue)
class VenueAdmin(TranslatableAdmin):
    list_display = ("id", "name", "created_at", "updated_at")
    list_display_links = ("id", "name")
    exclude = ("id", "created_at", "updated_at")
