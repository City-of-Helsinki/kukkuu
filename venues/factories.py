import factory

from projects.models import Project
from venues.models import Venue


class VenueFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("text", max_nb_chars=64)
    description = factory.Faker("text")
    address = factory.Faker("address")
    accessibility_info = factory.Faker("text")
    arrival_instructions = factory.Faker("text")
    additional_info = factory.Faker("text")
    wc_and_facilities = factory.Faker("text")
    www_url = factory.Faker("url")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))

    class Meta:
        model = Venue
        skip_postgeneration_save = True  # Not needed after factory v4.0.0
