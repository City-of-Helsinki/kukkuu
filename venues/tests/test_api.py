from copy import deepcopy

import pytest
from graphql_relay import to_global_id

from common.tests.utils import assert_permission_denied
from venues.models import Venue


@pytest.fixture(autouse=True)
def autouse_db(db):
    pass


VENUES_QUERY = """
query Venues {
  venues {
    edges {
      node {
        translations {
          name
          description
          address
          accessibilityInfo
          arrivalInstructions
          additionalInfo
          wwwUrl
          languageCode
        }
        occurrences {
          edges {
            node {
              time
              event {
                translations {
                    name
                    shortDescription
                    description
                    languageCode
                }
                  image
                  participantsPerInvite
                  capacityPerOccurrence
                  publishedAt
                  duration
              }
            }
          }
        }
      }
    }
  }
}
"""

VENUE_QUERY = """
query Venue($id: ID!) {
  venue(id: $id){
    translations{
        name
        description
        address
        accessibilityInfo
        arrivalInstructions
        additionalInfo
        wwwUrl
        languageCode
    }
    occurrences{
      edges{
        node{
          time
          event {
              translations {
                name
                shortDescription
                description
                languageCode
              }
              image
              participantsPerInvite
              capacityPerOccurrence
              publishedAt
              duration
            }
        }
      }
    }
  }
}
"""

ADD_VENUE_MUTATION = """
mutation AddVenue($input: AddVenueMutationInput!) {
  addVenue(input: $input) {
    venue {
      translations{
        languageCode
        name
        description
        address
        accessibilityInfo
        arrivalInstructions
        additionalInfo
        wwwUrl
      }
    }
  }
}
"""

ADD_VENUE_VARIABLES = {
    "input": {
        "translations": [
            {
                "name": "Venue name",
                "description": "Venue description",
                "languageCode": "fi",
                "address": "Address",
                "accessibilityInfo": "Accessibility info",
                "arrivalInstructions": "Arrival instruction",
                "additionalInfo": "Additional info",
                "wwwUrl": "www.url.com",
            }
        ],
    }
}

UPDATE_VENUE_MUTATION = """
mutation updateVenue($input: UpdateVenueMutationInput!) {
  updateVenue(input: $input) {
    venue {
      id
      translations{
        languageCode
        name
        description
        address
        accessibilityInfo
        arrivalInstructions
        additionalInfo
        wwwUrl
      }
    }
  }
}
"""

UPDATE_VENUE_VARIABLES = {
    "input": {
        "id": "",
        "translations": [
            {
                "name": "Venue name",
                "description": "Venue description",
                "languageCode": "fi",
                "address": "Address",
                "accessibilityInfo": "Accessibility info",
                "arrivalInstructions": "Arrival instruction",
                "additionalInfo": "Additional info",
                "wwwUrl": "www.url.com",
            }
        ],
    }
}

DELETE_VENUE_MUTATION = """
mutation DeleteVenue($input: DeleteVenueMutationInput!) {
  deleteVenue(input: $input) {
    __typename
  }
}
"""


def test_venues_query_unauthenticated(api_client):
    executed = api_client.execute(VENUES_QUERY)

    assert_permission_denied(executed)


def test_venues_query_normal_user(snapshot, user_api_client, occurrence):
    executed = user_api_client.execute(VENUES_QUERY)

    snapshot.assert_match(executed)


def test_venue_query_unauthenticated(api_client, occurrence):
    venue = occurrence.venue
    variables = {"id": to_global_id("VenueNode", venue.id)}
    executed = api_client.execute(VENUE_QUERY, variables=variables)

    assert_permission_denied(executed)


def test_venue_query_normal_user(snapshot, user_api_client, occurrence):
    venue = occurrence.venue
    variables = {"id": to_global_id("VenueNode", venue.id)}
    executed = user_api_client.execute(VENUE_QUERY, variables=variables)

    snapshot.assert_match(executed)


def test_add_venue_permission_denied(api_client, user_api_client):
    executed = api_client.execute(ADD_VENUE_MUTATION, variables=ADD_VENUE_VARIABLES)
    assert_permission_denied(executed)

    executed = user_api_client.execute(
        ADD_VENUE_MUTATION, variables=ADD_VENUE_VARIABLES
    )
    assert_permission_denied(executed)


def test_add_venue_staff_user(snapshot, staff_api_client):
    executed = staff_api_client.execute(
        ADD_VENUE_MUTATION, variables=ADD_VENUE_VARIABLES
    )
    snapshot.assert_match(executed)


def test_update_venue_permission_denied(api_client, user_api_client):
    executed = api_client.execute(
        UPDATE_VENUE_MUTATION, variables=UPDATE_VENUE_VARIABLES
    )
    assert_permission_denied(executed)

    executed = user_api_client.execute(
        UPDATE_VENUE_MUTATION, variables=UPDATE_VENUE_VARIABLES
    )
    assert_permission_denied(executed)


def test_update_venue_staff_user(snapshot, staff_api_client, venue):
    venue_variables = deepcopy(UPDATE_VENUE_VARIABLES)
    venue_variables["input"]["id"] = to_global_id("VenueNode", venue.id)
    executed = staff_api_client.execute(
        UPDATE_VENUE_MUTATION, variables=venue_variables
    )
    snapshot.assert_match(executed)


def test_delete_venue_permission_denied(api_client, user_api_client):
    executed = api_client.execute(
        DELETE_VENUE_MUTATION, variables={"input": {"id": ""}}
    )
    assert_permission_denied(executed)

    executed = user_api_client.execute(
        DELETE_VENUE_MUTATION, variables={"input": {"id": ""}}
    )
    assert_permission_denied(executed)


def test_delete_venue_staff_user(staff_api_client, venue):
    staff_api_client.execute(
        DELETE_VENUE_MUTATION,
        variables={"input": {"id": to_global_id("VenueNode", venue.id)}},
    )
    assert Venue.objects.count() == 0
