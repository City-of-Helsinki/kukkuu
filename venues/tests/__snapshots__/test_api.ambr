# serializer version: 1
# name: test_add_venue_project_user
  dict({
    'data': dict({
      'addVenue': dict({
        'venue': dict({
          'project': dict({
            'year': 2020,
          }),
          'translations': list([
            dict({
              'accessibilityInfo': 'Accessibility info',
              'additionalInfo': 'Additional info',
              'address': 'Address',
              'arrivalInstructions': 'Arrival instruction',
              'description': 'Venue description',
              'languageCode': 'FI',
              'name': 'Venue name',
              'wcAndFacilities': 'WC & Facilities',
              'wwwUrl': 'www.url.com',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_update_venue_project_user
  dict({
    'data': dict({
      'updateVenue': dict({
        'venue': dict({
          'project': dict({
            'year': 2020,
          }),
          'translations': list([
            dict({
              'accessibilityInfo': 'Accessibility info',
              'additionalInfo': 'Additional info',
              'address': 'Address',
              'arrivalInstructions': 'Arrival instruction',
              'description': 'Venue description',
              'languageCode': 'FI',
              'name': 'Venue name',
              'wcAndFacilities': 'WC & Facilities',
              'wwwUrl': 'www.url.com',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_venue_query_normal_user
  dict({
    'data': dict({
      'venue': dict({
        'accessibilityInfo': '''
          Like lay still bar. From daughter order stay sign discover eight.
          Scientist service wonder everything pay. Moment strong hand push book and interesting sit.
        ''',
        'additionalInfo': 'Training thought price. Effort clear and local challenge box. Care figure mention wrong when lead involve.',
        'address': '''
          04883 Mary Corner
          Port Mikeview, NY 31053
        ''',
        'arrivalInstructions': 'Benefit treat final central. Past ready join enjoy. Huge get this success commercial recently from.',
        'description': '''
          Together history perform. Respond draw military dog hospital number. Certainly again thought summer because serious listen.
          Page box child care any concern. Defense level church use.
        ''',
        'name': 'Poor lawyer treat free heart significant.',
        'occurrences': dict({
          'edges': list([
          ]),
        }),
        'project': dict({
          'year': 2020,
        }),
        'translations': list([
          dict({
            'accessibilityInfo': '''
              Like lay still bar. From daughter order stay sign discover eight.
              Scientist service wonder everything pay. Moment strong hand push book and interesting sit.
            ''',
            'additionalInfo': 'Training thought price. Effort clear and local challenge box. Care figure mention wrong when lead involve.',
            'address': '''
              04883 Mary Corner
              Port Mikeview, NY 31053
            ''',
            'arrivalInstructions': 'Benefit treat final central. Past ready join enjoy. Huge get this success commercial recently from.',
            'description': '''
              Together history perform. Respond draw military dog hospital number. Certainly again thought summer because serious listen.
              Page box child care any concern. Defense level church use.
            ''',
            'languageCode': 'FI',
            'name': 'Poor lawyer treat free heart significant.',
            'wcAndFacilities': '''
              State social believe policy. Score think turn argue present.
              Prevent pressure point. Voice radio happen color scene.
              Assume training seek full several. Authority develop identify ready.
            ''',
            'wwwUrl': 'http://www.brooks.com/',
          }),
        ]),
        'wcAndFacilities': '''
          State social believe policy. Score think turn argue present.
          Prevent pressure point. Voice radio happen color scene.
          Assume training seek full several. Authority develop identify ready.
        ''',
        'wwwUrl': 'http://www.brooks.com/',
      }),
    }),
  })
# ---
# name: test_venues_project_filter
  dict({
    'data': dict({
      'venues': dict({
        'edges': list([
          dict({
            'node': dict({
              'name': 'Should be returned',
            }),
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_venues_query_normal_user
  dict({
    'data': dict({
      'venues': dict({
        'edges': list([
          dict({
            'node': dict({
              'accessibilityInfo': '''
                Like lay still bar. From daughter order stay sign discover eight.
                Scientist service wonder everything pay. Moment strong hand push book and interesting sit.
              ''',
              'additionalInfo': 'Training thought price. Effort clear and local challenge box. Care figure mention wrong when lead involve.',
              'address': '''
                04883 Mary Corner
                Port Mikeview, NY 31053
              ''',
              'arrivalInstructions': 'Benefit treat final central. Past ready join enjoy. Huge get this success commercial recently from.',
              'description': '''
                Together history perform. Respond draw military dog hospital number. Certainly again thought summer because serious listen.
                Page box child care any concern. Defense level church use.
              ''',
              'name': 'Poor lawyer treat free heart significant.',
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'translations': list([
                dict({
                  'accessibilityInfo': '''
                    Like lay still bar. From daughter order stay sign discover eight.
                    Scientist service wonder everything pay. Moment strong hand push book and interesting sit.
                  ''',
                  'additionalInfo': 'Training thought price. Effort clear and local challenge box. Care figure mention wrong when lead involve.',
                  'address': '''
                    04883 Mary Corner
                    Port Mikeview, NY 31053
                  ''',
                  'arrivalInstructions': 'Benefit treat final central. Past ready join enjoy. Huge get this success commercial recently from.',
                  'description': '''
                    Together history perform. Respond draw military dog hospital number. Certainly again thought summer because serious listen.
                    Page box child care any concern. Defense level church use.
                  ''',
                  'languageCode': 'FI',
                  'name': 'Poor lawyer treat free heart significant.',
                  'wwwUrl': 'http://www.brooks.com/',
                }),
              ]),
              'wcAndFacilities': '''
                State social believe policy. Score think turn argue present.
                Prevent pressure point. Voice radio happen color scene.
                Assume training seek full several. Authority develop identify ready.
              ''',
              'wwwUrl': 'http://www.brooks.com/',
            }),
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_venues_query_ordering[en]
  dict({
    'data': dict({
      'venues': dict({
        'edges': list([
          dict({
            'node': dict({
              'name': '2 in Finnish',
            }),
          }),
          dict({
            'node': dict({
              'name': '1 in English',
            }),
          }),
          dict({
            'node': dict({
              'name': '4 in Finnish',
            }),
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_venues_query_ordering[fi]
  dict({
    'data': dict({
      'venues': dict({
        'edges': list([
          dict({
            'node': dict({
              'name': '2 in Finnish',
            }),
          }),
          dict({
            'node': dict({
              'name': '3 in Finnish',
            }),
          }),
          dict({
            'node': dict({
              'name': '4 in Finnish',
            }),
          }),
        ]),
      }),
    }),
  })
# ---
